narwhal/primary/src/header_waiter.rs

```diff
@@ -305,27 +322,20 @@ impl HeaderWaiter {
                 },
 
                 // we poll the availability of a slot to send the result to the core simultaneously
-                Some(header) = waiting.next() => {
-                    let header = match header{
+                Some(result) = waiting.next() => {
+                    match result{
                         Err(err) => {
-                            warn!("Error fetching header {}", err);
-                            continue;
+                            error!("Error loading header's dependencies {:?}", err);
+
+                            self.cleanup_pending_requests(&err.header());
+                        },
+                        Ok(header) => {
+                            self.cleanup_pending_requests(&header);
+
+                            // Ok to drop the header if core is overloaded.
+                            let _ = self.tx_headers_loopback.try_send(header);
```

Comment:
When we used to have `retries` in `header_waiter` it was ok to not send the header back to core - and do it only in the success scenario (like here). I believe though this now could create under some circumstances a loss of liveness. I keep the behaviour as is and not re-send header back to Core in case of error so can avoid any infinite loop that could be caused here without necessary backoff. As I said on the description @aschran will completely remove this component so don't want to spend more effort now